### ISSUE TITLE
Deno Module

### ISSUE THREAD

**Issue by felipefdl** [knex/knex#3851]
# Feature discussion / request

I would like to use Knex on Deno (https://deno.land/).

**Comment by kibertoad** [comment]
@felipefdl Are there any solutions for CommonJS and npm dependency support available for Deno already?

**Comment by felipefdl** [comment]
@kibertoad Not sure, probably not. Maybe it's not time to implement yet, but will be great have knex on Deno.

**Comment by kibertoad** [comment]
@felipefdl Definitely. It's not clear what needs to be done for that, as obviously full fork of Knex.js over to Deno is not really feasible or sustainable.
What changes in Knex codebase would make it more compatible with Deno without breaking Node.js support?

**Comment by felipefdl** [comment]
The major problem is on import dependencies since Deno doesn't use Commonjs and doesn't use package.json as well. If Knex was written typescript, maybe it can use an alias on imports, but it isn't the case, you can use a trick, creating a file with all dependencies used by the node, and another file with all dependency used by Deno, and import by the platform. I don't know if it is a good design, but it is a point to think about.

**Comment by felipefdl** [comment]
Also, the Deno has (in construction) module to add Node Compatibility.
https://deno.land/std/node/

Maybe when it is done, we can run knex without problem.

**Comment by shellscape** [comment]
Stopping by to lend my support for this request.

**Comment by craigcosmo** [comment]
I want this too

**Comment by pasha-bolokhov** [comment]
I want this three

**Comment by sunw31** [comment]
I want this four

**Comment by kibertoad** [comment]
@OlivierCavadenti What do you think about conversion of knex to TS for 3.0?

**Comment by OlivierCavadenti** [comment]
> @OlivierCavadenti What do you think about conversion of knex to TS for 3.0?

Yeah sure its a good candidate, i can take a look of this issue after that

**Comment by kibertoad** [comment]
#5144 is a good entry point for that, actually

**Comment by code-ape** [comment]
Friendly reminder that Deno has different "primitives" (i.e. crypto, network, buffers, etc) than Node.js so presumably the Node.js SQL drivers won't work with it. Especially those that build `.node` C++ modules. Here's the Deno Postgres driver for reference: https://github.com/denodrivers/postgres

**Comment by pasha-bolokhov** [comment]
Why not use the Deno drivers then?

**Comment by vanpipy** [comment]
Ok, I want now, what is going on here?

**Comment by kibertoad** [comment]
Nothing. Feel free to send PRs :D

**Comment by vanpipy** [comment]
> Nothing. Feel free to send PRs :D

Sadness, I will try but do not promise...:D

**Comment by vanpipy** [comment]
I found a way to use the `knex` without any changed, that is importing the module via an online CDN site like [esm.sh](https://esm.sh/), but the question is coming. All of the CDN sites is just giving the module used in the browser environment, this means a module should a cross-platform module always, or the dev needs polyfill the gap to reach the `cross-platform` goal.

An example, trying to use the `knex` via [Skypack](https://cdn.skypack.dev/knex@2.2.0) (yep, i tried...), an error throwed as below:

```
error: Uncaught (in promise) Error: [Package Error] "timers" does not exist. (Imported by "tarn").
```

Because of the `tarn` is not a `'cross-platform'` module, the `timers` cannot be found via the CDN. I just think to make every module is a `cross-platform` or try to polyfill everything is not good idea. Or better?

Anyway, importing module from the CDN is great way and the problem will be solved if the CND can give the module base different environment.

**Comment by felipefdl** [comment]
The Deno 1.25 added npm support, maybe it works, i didn’t try yet.

 https://deno.com/blog/v1.25

**Comment by vanpipy** [comment]
> The Deno 1.25 added npm support, maybe it works, i didn’t try yet.
> 
> https://deno.com/blog/v1.25

Thanks, it looks great. I am trying!

**Comment by vanpipy** [comment]
Hi, @felipefdl ,

For now, the `deno` cannot support to use node package fully. Maybe `--unstable` to stable is the day. Here is the [detail](https://github.com/denoland/deno/issues/15717).

**Comment by vanpipy** [comment]
Run with deno 1.26.1 but failed, has anyone tried it yet?

**Comment by jacoborus** [comment]
It looks like the problem is the peer dependencies, I believe it could be solved by allowing to manually pass the DB driver in the config. Something like:

```
import k from "npm:knex";
import sqlite3 from "npm:sqlite3";

const knex = k({
  client: "sqlite3",
  driver: sqlite3,
  connection: {
    filename: "./data.db",
  },
});
```

**Comment by mhio** [comment]
> It looks like the problem is the peer dependencies, I believe it could be solved by allowing to manually pass the DB driver in the config. Something like:

@jacoborus  I don't think deno will support any native node modules like sqlite3/better-sqlite3 as they are compiled against specific node versions. 

If knex does port across ok as is, it would probably need dialects for deno's [sqlite](https://deno.land/x/sqlite)/[sqlite3](https://deno.land/x/sqlite3) or waiting for the existing dialects to publish deno FFI versions. 

The pure js drivers like `pg` might work in your suggested format (but there's been no official work on [making deno supported](https://github.com/brianc/node-postgres/issues/2225))

**Comment by LeulAria** [comment]
++++

**Comment by LeulAria** [comment]
++++ in bun
